fe1-web/src/features/witness/components/WitnessNotification.tsx

```diff
@@ -0,0 +1,106 @@
+import PropTypes from 'prop-types';
+import React, { FunctionComponent, useEffect, useMemo } from 'react';
+import { Text, View } from 'react-native';
+import { Button } from 'react-native-elements';
+import { useSelector } from 'react-redux';
+
+import { dispatch } from 'core/redux';
+
+import { WitnessHooks } from '../hooks';
+import { WitnessFeature, MESSAGE_TO_WITNESS_NOTIFICATION_TYPE } from '../interface';
+import { requestWitnessMessage } from '../network/WitnessMessageApi';
+import { makeMessageToWitnessSelector, removeMessageToWitness } from '../reducer';
+
+const WitnessNotification = ({ notification, navigateToNotificationScreen }: IPropTypes) => {
+  const messageSelector = useMemo(
+    () => makeMessageToWitnessSelector(notification.messageId),
+    [notification.messageId],
+  );
+  const message = useSelector(messageSelector);
+
+  const discardNotifications = WitnessHooks.useDiscardNotifications();
+  const markNotificationAsRead = WitnessHooks.useMarkNotificationAsRead();
+  const isEnabled = WitnessHooks.useIsEnabled();
+  const laoId = WitnessHooks.useCurrentLaoId();
+
+  const onWitness = () => {
+    if (message) {
+      dispatch(
+        discardNotifications({ laoId: laoId.valueOf(), notificationIds: [notification.id] }),
+      );
+      if (isEnabled) {
+        requestWitnessMessage(message.channel, message.message_id);
+      }
+      dispatch(removeMessageToWitness(message.message_id.valueOf()));
+      navigateToNotificationScreen();
+    }
+  };
+
+  const onDecline = () => {
+    if (message) {
+      dispatch(markNotificationAsRead({ laoId: laoId.valueOf(), notificationId: notification.id }));
+      navigateToNotificationScreen();
+    }
+  };
+
+  // if the notification state somehow gets out of sync, remove the corresponding notification
+  useEffect(() => {
+    if (!message) {
+      dispatch(
+        discardNotifications({ laoId: laoId.valueOf(), notificationIds: [notification.id] }),
+      );
+      navigateToNotificationScreen();
+    }
+  }, [laoId, navigateToNotificationScreen, discardNotifications, notification.id, message]);
+
+  return (
+    <View>
+      <Text>{JSON.stringify(notification)}</Text>
+      <Text>{JSON.stringify(message)}</Text>
+      <Button title="Witness Message" onPress={onWitness} />
+      <Button title="Decline Message" onPress={onDecline} />
+    </View>
+  );
+};
+
+const propTypes = {
+  notification: PropTypes.shape({
+    // WitnessFeature.MessageToWitnessNotification
+    id: PropTypes.number.isRequired,
+    timestamp: PropTypes.number.isRequired,
+    title: PropTypes.string.isRequired,
+    type: PropTypes.string.isRequired,
+    messageId: PropTypes.string.isRequired,
+  }).isRequired,
```

Comment:
Couldn't it be treated as an object instance? I mean, if an object contains all those fields, you can just write `Prototypes.instanceOf(Object)`.
EDIT: It seems to match the `Notification` interface.

Reply:
Ah yes for sure, that is horrible! I am just not sure whether it is a good idea to add a class just for the sake of being able to validate it here using PropTypes 🤔 But maybe it is actually worth it, what do you think? Maybe there will be some other use for this class at a later point in time 😂

Reply:
Another way of handling this could be to add this `PropTypes.shape` in [this file](https://github.com/dedis/popstellar/blob/master/fe1-web/src/resources/Props.ts). The problem is that it would again need to be updated when the notification implementation changes (as Pierluca said), but this time we're sure that it will be updated only once.
Perhaps a class could be worth it to ensure this validation, yes.

Reply:
At the moment I am not sure whether a class is a good idea 🤔 I mean it would work but it I think it would require passing the constructor of a class:

The `WitnessNotification.tsx` component is passed to the notification feature and it is used of a given notification is of type `WitnessNotification`. We could define a class for `WitnessNotification` that extends some `BasicNotification` class defined in the `notification` feature so that the witness feature knows the exact type and the notification feature still has some idea of the datatypes it is operating on.

But then in order to extend `BasicNotification` in the `witness` feature we would have to pass this via dependency injection to the witness feature. This should work but I am not 100% how intuitive it is. Do you see a better solution or do you think this is intuitive?